File: src/main/java/org/shingas/viewerSMP/events/CosmeticApply.java
package org.shingas.viewerSMP.events;

import net.kyori.adventure.text.Component;
import org.bukkit.Material;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.Damageable;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.components.CustomModelDataComponent;
import org.shingas.viewerSMP.ViewerSMP;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class CosmeticApply implements Listener {

    private final FileConfiguration config = ViewerSMP.getConfigManager().cosmetics();

    public boolean isItemApplicable(String mat, int token) {
        String path = "CosmeticTokens.Token" + token + ".type";
        List<String> types = config.getStringList(path);
        for (String type : types) {
            if (mat.contains(type)) return true;
        }
        return false;
    }

    public int isCosmeticToken(ItemStack item) {
        if (item.getType() == Material.PAPER) {
            ItemMeta itemMeta = item.getItemMeta();
            if (itemMeta.hasCustomModelDataComponent()) {
                int modelData = itemMeta.getCustomModelDataComponent().getFloats().getFirst().intValue();
                String path = "CosmeticTokens.Token" + modelData + ".modelData";
                if(config.isSet(path)) {
                    return config.getInt(path);
                }
            }
        }
        return 0;
    }

    @EventHandler
    public void onInventoryClick(InventoryClickEvent event) {
        Player player = (Player) event.getWhoClicked();
        ItemStack item = event.getCurrentItem();
        ItemStack tokenItem = event.getCursor();

        if (item != null) {
            int token = isCosmeticToken(tokenItem);
            if (token != 0) {

                if (!isItemApplicable(item.getType().toString().toLowerCase(), token+1)) return;

                //Create the Item/ItemMetas
                ItemMeta oldMeta = item.getItemMeta();
                ItemStack newItem = new ItemStack(item.getType(), 1);
                ItemMeta newMeta = newItem.getItemMeta();

                //Get Name and Enchants of old Sword
                Component oldName = oldMeta.displayName();
                Map<Enchantment, Integer> oldEnchants = item.getEnchantments();


                //Get Damage and Apply to new Item
                if (oldMeta instanceof Damageable oldDamageable && newMeta instanceof Damageable newDamageable) {
                    int damage = oldDamageable.getDamage();
                    newDamageable.setDamage(damage);
                    newItem.setItemMeta(newMeta);
                }

                //Apply the custom model data to the new sword

                CustomModelDataComponent cmd = newMeta.getCustomModelDataComponent();
                List<Float> floats = new ArrayList<>(cmd.getFloats());
                floats.add((float) token);
                cmd.setFloats(floats);
                newMeta.setCustomModelDataComponent(cmd);

                //Apply the enchants to new sword
                for (Map.Entry<Enchantment, Integer> entry : oldEnchants.entrySet()) {
                    Enchantment enchant = entry.getKey();
                    int level = entry.getValue();
                    newMeta.addEnchant(enchant, level, true);
                }

                //Apply the name to new sword
                newMeta.displayName(oldName);


                //Apply the meta with all the changes and give item to player
                newItem.setItemMeta(newMeta);

                event.setCancelled(true);
                event.setCurrentItem(new ItemStack(Material.AIR));
                event.getView().getCursor().setAmount(0);
                player.getInventory().addItem(newItem);
            }
        }

    }

}

File: src/main/kotlin/com/shingaspt/viewerSMP/events/CosmeticApply.kt
package com.shingaspt.viewerSMP.events

import com.shingaspt.viewerSMP.ViewerSMP
import io.papermc.paper.registry.data.dialog.body.DialogBody.item
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.components.CustomModelDataComponent


class CosmeticApply : Listener {

    private val swords = listOf(
        Material.WOODEN_SWORD,
        Material.STONE_SWORD,
        Material.COPPER_SWORD,
        Material.IRON_SWORD,
        Material.GOLDEN_SWORD,
        Material.DIAMOND_SWORD,
        Material.NETHERITE_SWORD,
    )

    private val config = ViewerSMP.instance.configManager

    fun isSword(type: Material): Boolean {
        return swords.contains(type)
    }

    fun isCosmeticToken(item: ItemStack) : Int {
        if (item.type == Material.PAPER) {
            val itemMeta = item.itemMeta
            if (itemMeta.hasCustomModelDataComponent()) {
                val modelData = itemMeta.customModelDataComponent.floats.first().toInt()
                val path = "CosmeticTokens.Token$modelData.ModelData"
                if(config.pathExists(path)) {
                    return config.getInt(path)
                }
            }
        }
        return 0
    }

    @EventHandler
    fun onInventoryClick(event: InventoryClickEvent) {
        val player = event.whoClicked as Player
        val swordItem = event.currentItem
        val tokenItem = event.cursor

        if (swordItem != null) {
            if (isSword(swordItem.type)) {
                val token = isCosmeticToken(tokenItem)
                if (token != 0) {
                    //Get Name and Enchants of old Sword
                    val swordName = swordItem.itemMeta.displayName()
                    val swordEnchants = swordItem.enchantments

                    //Create item stack of new sword
                    val newSword = ItemStack(swordItem.type, 1)
                    val swordItemMeta = newSword.itemMeta

                    //Apply the custom model data to the new sword
                    val cmd = swordItemMeta.customModelDataComponent
                    cmd.floats = listOf(token.toFloat())
                    swordItemMeta.setCustomModelDataComponent(cmd)

                    //Apply the enchants to new sword
                    for ((enchant, level) in swordEnchants) {
                        swordItemMeta.addEnchant(enchant, level, true) // true = ignore level cap
                    }

                    //Apply the name to new sword
                    swordItemMeta.displayName(swordName)

                    //Apply the meta with all the changes and give item to player
                    newSword.itemMeta = swordItemMeta

                    event.isCancelled = true
                    event.currentItem = null
                    event.view.cursor.let { event.view.setCursor(null) }

                    player.inventory.addItem(newSword)
                }
            }
        }
    }

}